src/database.rs

```diff
@@ -0,0 +1,132 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+use ethers_core::types::BlockId;
+use ethers_providers::{Http, Provider};
+use pyo3::{PyErr, PyResult};
+use revm::Database;
+use revm::db::{CacheDB, DbAccount, EthersDB};
+use revm::precompile::{Address, B256};
+use revm::primitives::{AccountInfo, Bytecode, HashMap, State};
+use revm_interpreter::primitives::db::{DatabaseCommit, DatabaseRef};
+use ruint::aliases::U256;
+
+use crate::empty_db_wrapper::EmptyDBWrapper;
+use crate::utils::pyerr;
+
+type MemDB = CacheDB<EmptyDBWrapper>;
+type ForkDB = CacheDB<EthersDB<Provider<Http>>>;
+
+
+/// A wrapper around the `CacheDB` and `EthersDB` to provide a common interface
+/// without needing dynamic lifetime and generic parameters (unsupported in PyO3)
+#[derive(Clone, Debug)]
+pub(crate) enum DB {
+    Memory(MemDB),
+    Fork(ForkDB),
+}
+
+impl DB {
+    pub(crate) fn new_memory() -> Self {
+        DB::Memory(MemDB::new(EmptyDBWrapper::default()))
+    }
+
+    pub(crate) fn new_fork(
+        fork_url: &str,
+        fork_block_number: Option<&str>,
+    ) -> PyResult<Self> {
+        let provider = Provider::<Http>::try_from(fork_url).map_err(pyerr)?;
+        let block = fork_block_number.map(|n| BlockId::from_str(n)).map_or(Ok(None), |v| v.map(Some)).map_err(pyerr)?;
+        let db = EthersDB::new(Arc::new(provider), block).unwrap();
+        // todo: do we need to get the blockEnv from the client?
```

Comment:
Something like this is maybe needed for fetch block: https://github.com/bluealloy/revm/blob/main/crates/revm/src/db/ethersdb.rs#L23

Reply:
`block_number` is private on ethersdb, I don't think we need to retrieve it here again